core/raftlease/fsm.go

```diff
@@ -101,35 +107,46 @@ func (f *FSM) extend(key lease.Key, holder string, duration time.Duration) *resp
 	return &response{}
 }
 
+func (f *FSM) pin(key lease.Key) *response {
+	f.pinned[key] = true
+	return &response{}
+}
+
+func (f *FSM) unpin(key lease.Key) *response {
+	delete(f.pinned, key)
+	return &response{}
+}
+
 func (f *FSM) setTime(oldTime, newTime time.Time) *response {
-	f.mu.Lock()
-	defer f.mu.Unlock()
 	if f.globalTime != oldTime {
 		return &response{err: globalclock.ErrConcurrentUpdate}
 	}
 	f.globalTime = newTime
+	return &response{expired: f.expired(newTime)}
+}
+
+// expired returns a collection of keys for leases that have expired.
+// Any pinned leases are not included in the return.
+func (f *FSM) expired(newTime time.Time) []lease.Key {
 	var expired []lease.Key
 	for key, entry := range f.entries {
 		expiry := entry.start.Add(entry.duration)
-		if expiry.Before(newTime) {
+		if expiry.Before(newTime) && !f.pinned[key] {
 			delete(f.entries, key)
 			expired = append(expired, key)
 		}
 	}
-	return &response{expired: expired}
+	return expired
 }
 
 // GlobalTime returns the FSM's internal time.
 func (f *FSM) GlobalTime() time.Time {
-	f.mu.Lock()
-	defer f.mu.Unlock()
 	return f.globalTime
 }
 
 // Leases gets information about all of the leases in the system.
 func (f *FSM) Leases(localTime time.Time) map[lease.Key]lease.Info {
 	f.mu.Lock()
-	defer f.mu.Unlock()
 	results := make(map[lease.Key]lease.Info)
 	for key, entry := range f.entries {
```

Comment:
We might find that the lease manager gets into a spin if we return leases that look like they should be expired but then can't be expired using the store method. It's probably better to report pinned leases with an expiry in the future to avoid that.

Reply:
Done.